### PR TITLE
Use boost implementation of legendre polynomials instead of gsl.

### PR DESCRIPTION
**Pull request by marcfehling**
Noticed in #11254.

Blocked by #11286.

Instead of using the boost variant directly, we could also introduce a `legendre` function in the `std_cxx17` namespace which uses the boost implementation. This way we could remind ourselves of using the CXX17 function once we move to the more recent standard. I don't expect the results to be different, as they are also not different between boost and gsl.

What is your preference?

### CODE REVIEW

**Comment by masterleinad** [comment]
We already use `std::cyl_bessel_j*`. I don't see a reason not to do the same for the Legendre polynomials.

**Comment by bangerth** [comment]
Ah, yes, the hive mind works! I had forgotten about `std::cyl_bessel_j`.

@marcfehling -- can you add the same kind of approach to `include/deal.II/base/std_cxx17/cmath.h` and use that here?

**Comment by marcfehling** [comment]
The `std` functions use the type `unsigned int` for the index parameter, while the `boost` implementation uses `int`.

I provided them as functions with a `static_cast` instead of function pointers like with the Bessel functions. What do you think?

**Comment by marcfehling** [comment]
Blocked by #11286 as we include `cmath.h` twice now :-)

**Comment by marcfehling** [comment]
Grr, this issue also affects the double definition of the Legendre polynomials.
```
 /usr/bin/ld.gold: error: base/CMakeFiles/obj_base_release.dir/function_lib.cc.o: multiple definition of 'dealii::std_cxx17::legendre(unsigned int, double)'
```

I'll think of a fix.

**Comment by bangerth** [comment]
If you implement these as functions, then `constexpr` is the wrong approach. You want `inline` :-)

(As a matter of fact, we also want `inline` variables in #11286, but that's a C++17 feature...)

**Comment by marcfehling** [comment]
> If you implement these as functions, then `constexpr` is the wrong approach. You want `inline` :-)

Ah, since those are actual function and not just function pointers. I get it now :-)

**Review by bangerth — APPROVED** [review]

**Comment by bangerth** [comment]
Can you squash the last few commits?

**Comment by bangerth** [comment]
Blocked by #11286. Otherwise ok to merge once squashed.

**Comment by marcfehling** [comment]
Hmm it looks like the CI doesn't like the `constexpr` flag for the function pointers, as the linker complains particularly about the Bessel functions.
```
/usr/bin/ld.gold: error: base/CMakeFiles/obj_base_release.dir/function_lib.cc.o: multiple definition of 'dealii::std_cxx17::cyl_bessel_jl'
/usr/bin/ld.gold: fe/CMakeFiles/obj_fe_release.dir/fe_series_legendre.cc.o: previous definition here
/usr/bin/ld.gold: error: base/CMakeFiles/obj_base_release.dir/function_lib.cc.o: multiple definition of 'dealii::std_cxx17::cyl_bessel_jf'
/usr/bin/ld.gold: fe/CMakeFiles/obj_fe_release.dir/fe_series_legendre.cc.o: previous definition here
/usr/bin/ld.gold: error: base/CMakeFiles/obj_base_release.dir/function_lib.cc.o: multiple definition of 'dealii::std_cxx17::cyl_bessel_j'
/usr/bin/ld.gold: fe/CMakeFiles/obj_fe_release.dir/fe_series_legendre.cc.o: previous definition here
```

Shall we turn the function pointers to wrapper functions as well? Hmm, or maybe it would help to flag them static.

**Comment by bangerth** [comment]
Hm. Can you rebase to master and see again what happens?

**Review by masterleinad — COMMENTED** [review]
We should use a common style here. I would be fine with `inline` functions.

**Review by masterleinad — CHANGES_REQUESTED** [review]
We should use a common style here. I would be fine with inline functions.

**Comment by marcfehling** [comment]
The CI complains about the `constexpr` and `inline` flags for the function pointers. It works with a `static` flag or when turning them to wrapper functions.

I comply to @masterleinad's suggestion to use the `inline` flag in combination with wrapper functions.

**Review by masterleinad — APPROVED** [review]

**Comment by bangerth** [comment]
You'll need to run the indentation script.

**Comment by marcfehling** [comment]
Even the indentation script got stuck. I need to rebase now anyways. Give me a sec

**Comment by marcfehling** [comment]
I've changed the parameter names to be in accordance with this [`cmath.h` reference](https://en.cppreference.com/w/cpp/header/cmath).

Ready for merge.

**Review by bangerth — APPROVED** [review]

**Pull request merged**
